src/docs/tiered-storage.dox

```diff
@@ -0,0 +1,82 @@
+/*! @page tiered_storage Tiered Storage
+
+\warning The WiredTiger Tiered Storage support has not been developed, tested or tuned in
+production scenarios, and should be viewed as experimental code.
+
+@section tiered_background Background
+
+A tiered table is a single btree (possibly with multiple checkpoints) made
+up of blocks that reside in one or more files on the local file system and
+zero or more cloud objects (e.g., stored in S3). The btree spans this set of
+files and objects by encoding a block's location in the pointer that links
+blocks in the btree. A tiered btree is contrasted with a regular (non-tiered)
+btree that is contained in a single file. It is also different from an LSM
+(see @ref lsm) tree, which WiredTiger implements as multiple regular btrees,
+each in its own file.
+
+When in use, a tiered btree, like a regular btree, may have some
+recently used or modified data that resides in memory pages.
+This in-memory representation is the same between a tiered btree
+and a regular btree, it is only how data is stored on disk and in
+cloud objects that makes these btrees different.
+
+All writes (i.e., pages that are reconciled and written to storage by eviction
+or checkpoint) are written to a designated file, called the active file.
+All other files and objects that are part of the tiered table are read-only.
+Each object or file is given an object number, which appears as part of its
+name on disk, cloud and in metadata.
+
+When a tiered btree is created, there is a single file associated with it,
+it is given object number 1. This is the active file, and it only resides
+on disk. This is always true for an active file. When a file ceases
+to be an active file, writes to it will no longer be permitted.  At that
+point, such a file can be copied to cloud storage.  How this transition
+to a new active file occurs is the subject of the next section.
+
+@section flush_tier The flush_tier operation
+
+A fundamental operation for a tiered table is \a flush_tier, this occurs
+as a result of a call to WT_SESSION::checkpoint with a configuration
+string that includes \c "flush_tier=(enabled=true)".  Because flush_tier
+is a part of checkpoint, it is a system-wide operation. Thus, flush_tier
+allows the data for all tiered btrees to move from the local file system
+to object storage.
+
+The flush_tier checkpoint does all the operations normally associated with
+a checkpoint, and in addition:
+
+ -# identifies any tiered btrees that have changes since the previous checkpoint
+ -# for each such modified btree:
+     -# eviction is temporarily disabled on this btree, while waiting for any writes to the active file to drain
+     -# a new empty file is created named with the next object number, this will become the table's active file
+     -# switch the table's active file
+     -# enable eviction on this btree
+     -# queue the old active file to be written to object storage in the background
```

Comment:
See previous comment about this sequence.

Reply:
@keitharnoldsmith I think I missed the previous comment you're referring to?

Reply:
Hmm. I thought I talked about this elsewhere, but maybe I didn't save that comment...

A flush_tier checkpoint has a slightly different sequence of operations. In particular, eviction is only disabled when we are checkpointing the tree.

Checkpoint prepare phase: 
- Assemble list of all the dhandles that will be part of the checkpoint. For non-tiered tables this will be tables that have been modified since the last checkpoint. For tiered tables this will be tables modified since the last flush. See `checkpoint_flush_tier` called from `checkpoint_prepare`.
- For each tiered table, we create the new file, and queue a work item to flush the old one. The work item includes checkpoint generation information that is used by the tiered server thread to know when it is OK to flush the file to object storage (i.e., after the current checkpoint has completed). See `tiered_switch`, which is called from `checkpoint_flush_tier`.
- We do this work now so that it doesn't add time to the actual checkpoint when we have eviction disabled.

Data file checkpoint: 
- When we do the actual checkpoint for each table we block eviction to the table. So checkpoint is the only thread updating the table. This happens in `wt_sync_file`.
- At the end of the checkpoint, after writing the new checkpoint root but before re-enabling eviction, we switch the active file to use the new file created during the prepare phase. This ensures that everything that is part of the checkpoint is in the old file, and anything evicted after the checkpoint is in the next file.
- If the table is tiered, we fsync the old active file to ensure that all checkpoint updates are durable. (Side note: we don't have to do that here, we could do it after re-enabling eviction. But I'm just describing what the code does.)
- See `bm_checkpoint` for the active file switch and fsync.
- After switching the file we allow eviction again, as we finish `wt_sync_file`, and we checkpoint the next dhandle. 

Hope that makes sense. If not, feel free to ping me.